Add Installer documentation (#230)
vmware-tanzu/cluster-api-provider-bringyourownhost: docs/local_dev.md

docker rm -f $HOST_NAME
kind delete cluster
```

# Host Agent Installer

> The installer is responsible for detecting the BYOH OS, downloading a BYOH bundle and installing/uninstalling it.

Reviewer: ```suggestion
The installer is responsible for detecting the Host OS, downloading a BYOH bundle and installing/uninstalling it.
```

## Supported OS and Kubernetes
The current list of supported tuples of OS, kubernetes Version, BYOH Bundle Name can be retrieved with:
```shell
./cli --list-supported
```
An example output looks like:
<table>
    <tr>
        <td>OS</td>
        <td>K8S Version</td>
        <td>BYOH Bundle Name</td>
    </tr>
    <tr>
        <td>Ubuntu_20.04.*_x86-64</td>
        <td>v1.22.3</td>
        <td>byoh-bundle-ubuntu_20.04.1_x86-64_k8s_v1.22.3</td>
    </tr>
</table>
The '*' in OS means that all Ubuntu 20.04 patches will be handled by this BYOH bundle.

## Pre-requisites

> As of writing this, the following packages must be pre-installed on the BYOH host:

Reviewer: Question: These packages are only required for debian systems or other systems as well? Do we need to mention Ubuntu specifically?
```suggestion
As of writing this, the following packages must be pre-installed on the Host:
```

Author: As of now we support only Ubuntu. For RHEL these will be rpms

- socat
- ebtables
- ethtool
- conntrack
```shell
sudo apt-get install socat ebtables ethtool conntrack
```

## Creating a BYOH Bundle
### Kubernetes Ingredients
Optional. This step describes downloading kubernetes host components for Debian.
```shell
# Build docker image
(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)

# Create a directory for the ingredients and download to it

> (mkdir -p byoh-ingredients-download && docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients byoh-ingredients-deb)

Reviewer: Naive question: I was more aware of the usage `-v ${PWD}:/home/somepath` for volume mounts. But above is also okay if there are no downsides.

Author: I guess both are OK for an example.

docker rm -f $HOST_NAME
kind delete cluster
```

# Host Agent Installer
The installer is responsible for detecting the BYOH OS, downloading a BYOH bundle and installing/uninstalling it.

## Creating a BYOH Bundle

> ### Upstream Ingredients

Reviewer 2: ```suggestion
### Kubernetes Ingredients
```

Or some other better terminology than upstream

> Optional. This step describes downloading upstream k8s host components for Debian.

Reviewer 2: ```suggestion
Optional. This step describes downloading kubernetes host components for Debian.
```

```shell

> #Build docker image

Reviewer 2: ```suggestion
# Build docker image
```

(cd agent/installer/bundle_builder/ingredients/deb/ && docker build -t byoh-ingredients-deb .)

> #Create a directory for the ingredients and download to it

Reviewer 2: ```suggestion
# Create a directory for the ingredients and download to it
```
And all other occurrence of comments. Good to add a space for better readability

(mkdir -p byoh-ingredients-download && docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients byoh-ingredients-deb)
```

> ### Downstream Ingredients

Reviewer 2: ```suggestion
### Custom Ingredients
```

> This step describes providing custom k8s host components. They can be copied to to byoh-ingredients-download. Files must match the following globs:

Reviewer 2: ```suggestion
This step describes providing custom kubernetes host components. They can be copied to `byoh-ingredients-download`. Files must match the following globs:
```

```shell
*containerd*.tar
*kubeadm*.deb
*kubelet*.deb
*kubectl*.deb
*cri-tools*.deb
*kubernetes-cni*.deb
```

## Building a BYOH Bundle
```shell
#Build docker image

> (cd agent/installer/bundle_builder/ && docker build -t byoh-build-push-bundle .)

Reviewer 2: ```suggestion
cd agent/installer/bundle_builder/ && docker build -t byoh-build-push-bundle .
```

Author: Braces are needed so the current dir of the current shell is not changed and you can execute all commands in sequence. Braces start a subshell.

```

```shell

> #Build a BYOH bundle and publish it to an OCI-compliant repo

Reviewer 2: Also worth adding how to push to a local registry? wdyt?

Author: Pushing to local registry should be like pushing to non-local.

If there is something very special in pushing, like we need, you can use the BUILD_ONLY=0 which is already explained.

```

```shell
#Optionally, additional configuration can be included in the bundle by mounting a local path under /config of the container. It will be placed on top of any drop-in configuration created by the packages and tars in the bundle
docker run --rm -v `pwd`/byoh-ingredients-download:/ingredients -v`pwd`:/bundle -v`pwd`/agent/installer/bundle_builder/config/ubuntu/20_04/k8s/1_22 --env BUILD_ONLY=1 build-push-bundle
```

## CLI
The installer CLI exposes the installer package as a command line tool. For a list of all commands, run

```shell
./cli --help
```

Some examples:
```shell
#Will return if/how the current OS is detected
./cli --detect
```

```shell
#Will return tuples of OS, K8S Version, BYOH Bundle Name. Installer will use this bundle for these OS-es and k8s versions.
./cli --list-supported
```

```shell
#Will return the OS changes that installer will make during install and uninstall without actually doing them.
./cli --preview-os-changes --os Ubuntu_20.04.*_x86-64 --k8s v1.22.3
```

```shell
#Will detect the current OS and install BYOH bundle with k8s v1.22.3 from the default repo
sudo ./cli --install --k8s v1.22.3
```

> ```shell

Reviewer 2: Good to mention what the current default OS support is - either here or at the top in the Installer section.
P.S - yes, it needs to be updated everytime we add new support. I'm also okay having a support matrix in the top level `README`. Whatever you seem fit.

#Will override the OS detection and will use custom repo
sudo ./cli --install --os Ubuntu_20.04.1_x86-64 --bundle-repo 10.26.226.219:5000/repo --k8s v1.22.3

> ```

Reviewer 2: What about `uninstall`?
